File: tests/platform/ios_platform_registry.m
@import HuxerUIPlatform;

@interface HuxerUITestPlatformModule : NSObject <HUXPlatformModule>
@end

@implementation HuxerUITestPlatformModule

- (id<HUXPlatformCancellation>)invoke:(NSString*)method
                            arguments:(HUXPlatformPayload*)arguments
                               result:(id<HUXPlatformResult>)result {
  [result complete:[HUXPlatformPayload objectValue:@{
                    @"method" : [HUXPlatformPayload stringValue:method],
                    @"arguments" : arguments,
                  }]];
  return nil;
}

- (void)dispose {
}

@end

@interface HuxerUITestPlatformModuleFactory : NSObject <HUXUIKitPlatformModuleFactory>
@end

@implementation HuxerUITestPlatformModuleFactory

- (id<HUXPlatformModule>)createWithViewController:(UIViewController*)viewController
                                           options:(HUXPlatformPayload*)options
                                            events:(id<HUXPlatformEventEmitter>)events {
  (void)viewController;
  [events emit:@"created" payload:options];
  return [HuxerUITestPlatformModule new];
}

@end

@interface HuxerUITestPlatformView : NSObject <HUXUIKitPlatformView>
@property(nonatomic, readonly) UIView* view;
@end

@implementation HuxerUITestPlatformView

- (instancetype)init {
  self = [super init];
  if (self != nil) {
    _view = [UIView new];
  }
  return self;
}

- (void)dispose {
}

@end

@interface HuxerUITestPlatformViewFactory : NSObject <HUXUIKitPlatformViewFactory>
@end

@implementation HuxerUITestPlatformViewFactory

- (id<HUXUIKitPlatformView>)createWithViewController:(UIViewController*)viewController
                                           properties:(HUXPlatformPayload*)properties
                                               events:(id<HUXPlatformEventEmitter>)events {
  (void)viewController;
  [events emit:@"created" payload:properties];
  return [HuxerUITestPlatformView new];
}

@end

File: tests/platform/ios_platform_registry.swift
import HuxerUIPlatform
import UIKit

final class TestPlatformModule: NSObject, PlatformModule {
  func invoke(_ method: String, arguments: PlatformPayload, result: PlatformResult) -> PlatformCancellation? {
    result.complete(.object([
      "method": .string(method),
      "arguments": arguments,
    ]))
    return nil
  }

  func dispose() {}
}

final class TestPlatformModuleFactory: NSObject, UIKitPlatformModuleFactory {
  func create(with viewController: UIViewController, options: PlatformPayload,
              events: PlatformEventEmitter) -> PlatformModule {
    events.emit("created", payload: options)
    return TestPlatformModule()
  }
}

final class TestPlatformView: NSObject, UIKitPlatformView {
  let view = UIView()

  func dispose() {}
}

final class TestPlatformViewFactory: NSObject, UIKitPlatformViewFactory {
  func create(with viewController: UIViewController, properties: PlatformPayload,
              events: PlatformEventEmitter) -> UIKitPlatformView {
    events.emit("created", payload: properties)
    return TestPlatformView()
  }
}

let payload: PlatformPayload = .object([
  "enabled": .boolean(true),
  "count": .integer(3),
])
payload.validate(fields: ["enabled", "count"])

let textureSource = ExternalTextureSource(intrinsicSize: CGSize(width: 16, height: 9))
_ = textureSource.texture
textureSource.finish()
